flaky test: HTTPClientTests.testRaceNewRequestsVsShutdown

5.1 CI

```
Test Case 'HTTPClientTests.testRaceNewRequestsVsShutdown' started at 2020-05-21 11:37:53.333
/code/Tests/AsyncHTTPClientTests/HTTPClientTests.swift:1139: error: HTTPClientTests.testRaceNewRequestsVsShutdown : failed - Unexpected error: ProviderClosedError()
Assertion failed: file /code/Sources/AsyncHTTPClient/ConnectionsState.swift, line 92
Current stack trace:
0    libswiftCore.so                    0x00007ffaa5df032b <unavailable> + 3953451
1    libswiftCore.so                    0x00007ffaa5df03a0 swift_reportError + 50
2    libswiftCore.so                    0x00007ffaa5e61090 _swift_stdlib_reportFatalErrorInFile + 115
3    libswiftCore.so                    0x00007ffaa5d854fe <unavailable> + 3515646
4    libswiftCore.so                    0x00007ffaa5d85677 <unavailable> + 3516023
5    libswiftCore.so                    0x00007ffaa5b7addd <unavailable> + 1375709
6    libswiftCore.so                    0x00007ffaa5d5c458 <unavailable> + 3347544
7    libswiftCore.so                    0x00007ffaa5b7a179 <unavailable> + 1372537
8    async-http-clientPackageTests.xctest 0x000055d0f56dbaeb <unavailable> + 1346283
9    async-http-clientPackageTests.xctest 0x000055d0f56c8bd1 <unavailable> + 1268689
10   async-http-clientPackageTests.xctest 0x000055d0f56c8d04 <unavailable> + 1268996
11   libswiftCore.so                    0x00007ffaa5df1e8b <unavailable> + 3960459
12   async-http-clientPackageTests.xctest 0x000055d0f56c66a3 <unavailable> + 1259171
13   async-http-clientPackageTests.xctest 0x000055d0f56c6704 <unavailable> + 1259268
14   libswiftCore.so                    0x00007ffaa5df1e8b <unavailable> + 3960459
15   async-http-clientPackageTests.xctest 0x000055d0f56d5a24 <unavailable> + 1321508
16   async-http-clientPackageTests.xctest 0x000055d0f56d5dd3 <unavailable> + 1322451
17   libswiftCore.so                    0x00007ffaa5df1e8b <unavailable> + 3960459
18   async-http-clientPackageTests.xctest 0x000055d0f56d5e92 <unavailable> + 1322642
19   libswiftCore.so                    0x00007ffaa5df1e8b <unavailable> + 3960459
20   async-http-clientPackageTests.xctest 0x000055d0f5d910e2 <unavailable> + 8380642
21   libswiftCore.so                    0x00007ffaa5df1e8b <unavailable> + 3960459
22   async-http-clientPackageTests.xctest 0x000055d0f5d7f04f <unavailable> + 8306767
23   async-http-clientPackageTests.xctest 0x000055d0f5d825cb <unavailable> + 8320459
24   async-http-clientPackageTests.xctest 0x000055d0f5d823a5 <unavailable> + 8319909
25   async-http-clientPackageTests.xctest 0x000055d0f56f95c5 <unavailable> + 1467845
26   async-http-clientPackageTests.xctest 0x000055d0f5706cc8 <unavailable> + 1522888
27   async-http-clientPackageTests.xctest 0x000055d0f56cae18 <unavailable> + 1277464
28   async-http-clientPackageTests.xctest 0x000055d0f5706d61 <unavailable> + 1523041
29   async-http-clientPackageTests.xctest 0x000055d0f5d9b621 <unavailable> + 8422945
30   async-http-clientPackageTests.xctest 0x000055d0f5d9cd66 <unavailable> + 8428902
31   async-http-clientPackageTests.xctest 0x000055d0f5d8cf64 <unavailable> + 8363876
32   async-http-clientPackageTests.xctest 0x000055d0f5d91140 <unavailable> + 8380736
33   async-http-clientPackageTests.xctest 0x000055d0f5d878af <unavailable> + 8341679
34   async-http-clientPackageTests.xctest 0x000055d0f5d8bb23 <unavailable> + 8358691
35   async-http-clientPackageTests.xctest 0x000055d0f5d907f0 <unavailable> + 8378352
36   async-http-clientPackageTests.xctest 0x000055d0f5e419e8 <unavailable> + 9103848
37   async-http-clientPackageTests.xctest 0x000055d0f5e4961e <unavailable> + 9135646
38   async-http-clientPackageTests.xctest 0x000055d0f5e41a38 <unavailable> + 9103928
39   async-http-clientPackageTests.xctest 0x000055d0f5e46f9c <unavailable> + 9125788
40   async-http-clientPackageTests.xctest 0x000055d0f5e41ab5 <unavailable> + 9104053
41   async-http-clientPackageTests.xctest 0x000055d0f5e46e8c <unavailable> + 9125516
42   async-http-clientPackageTests.xctest 0x000055d0f56f0340 <unavailable> + 1430336
43   async-http-clientPackageTests.xctest 0x000055d0f5e46ef8 <unavailable> + 9125624
44   async-http-clientPackageTests.xctest 0x000055d0f5e38059 <unavailable> + 9064537
45   async-http-clientPackageTests.xctest 0x000055d0f5e3f5c4 <unavailable> + 9094596
46   async-http-clientPackageTests.xctest 0x000055d0f5d71780 <unavailable> + 8251264
47   async-http-clientPackageTests.xctest 0x000055d0f5d7257d <unavailable> + 8254845
48   async-http-clientPackageTests.xctest 0x000055d0f5d7e1d1 <unavailable> + 8303057
49   async-http-clientPackageTests.xctest 0x000055d0f5d72f5c <unavailable> + 8257372
50   async-http-clientPackageTests.xctest 0x000055d0f5eb6909 <unavailable> + 9582857
51   async-http-clientPackageTests.xctest 0x000055d0f5ebc34b <unavailable> + 9605963
52   async-http-clientPackageTests.xctest 0x000055d0f5ebc6b1 <unavailable> + 9606833
53   async-http-clientPackageTests.xctest 0x000055d0f5650003 <unavailable> + 774147
54   libpthread.so.0                    0x00007ffaa58156ba <unavailable> + 30394
55   libc.so.6                          0x00007ffaa39ff3b0 clone + 109
Exited with signal code 4
Build step 'Execute shell' marked build as failure
```

@artemredkin could this be a bug in the new pool implementation?

--- a/Sources/AsyncHTTPClient/ConnectionPool.swift
+++ b/Sources/AsyncHTTPClient/ConnectionPool.swift
@@ -489,8 +489,6 @@ class HTTP1ConnectionProvider {
 
     func close() -> EventLoopFuture<Bool> {
         if let (waiters, available, leased, clean) = self.lock.withLock({ self.state.close() }) {
-            print(waiters.count, available.count, leased.count, clean)
-
             waiters.forEach {
                 $0.promise.fail(HTTPClientError.cancelled)
             }

--- a/Sources/AsyncHTTPClient/ConnectionPool.swift
+++ b/Sources/AsyncHTTPClient/ConnectionPool.swift
@@ -80,12 +80,12 @@ final class ConnectionPool {
             if let existing = self.providers[key], existing.enqueue() {
                 return existing
             } else {
-                // Connection provider will be created with `pending = 1`
                 let provider = HTTP1ConnectionProvider(key: key,
                                                        eventLoop: taskEventLoop,
                                                        configuration: self.configuration,
                                                        pool: self,
                                                        backgroundActivityLogger: self.backgroundActivityLogger)
+                _ = provider.enqueue()
                 self.providers[key] = provider
                 return provider
             }
@@ -263,8 +263,6 @@ struct ConnectionKey: Hashable {
 /// of concurrent requests as it has built-in politeness regarding the maximum number
 /// of concurrent requests to the server.
 class HTTP1ConnectionProvider {
-    struct ProviderClosedError: Error {}
-
     /// The client configuration used to bootstrap new requests
     private let configuration: HTTPClient.Configuration
 

--- a/Sources/AsyncHTTPClient/ConnectionPool.swift
+++ b/Sources/AsyncHTTPClient/ConnectionPool.swift
@@ -85,7 +85,8 @@ final class ConnectionPool {
                                                        configuration: self.configuration,
                                                        pool: self,
                                                        backgroundActivityLogger: self.backgroundActivityLogger)
-                _ = provider.enqueue()
+                let enqueued = provider.enqueue()
+                assert(enqueued)
                 self.providers[key] = provider
                 return provider
             }
@@ -316,7 +317,7 @@ class HTTP1ConnectionProvider {
         self.state.assertInvariants()
     }
 
-    private func execute(_ action: Action, logger: Logger) {
+    func execute(_ action: Action, logger: Logger) {
         switch action {
         case .lease(let connection, let waiter):
             // if connection is became inactive, we create a new one.
@@ -488,8 +489,15 @@ class HTTP1ConnectionProvider {
 
     func close() -> EventLoopFuture<Bool> {
         if let (waiters, available, leased, clean) = self.lock.withLock({ self.state.close() }) {
+            print(waiters.count, available.count, leased.count, clean)
+
             waiters.forEach {
                 $0.promise.fail(HTTPClientError.cancelled)
+            }
+
+            if available.isEmpty, leased.isEmpty {
+                self.closePromise.succeed(())
+                return self.closePromise.futureResult.map { clean }
             }
 
             EventLoopFuture.andAllComplete(leased.map { $0.cancel() }, on: self.eventLoop).flatMap { _ in

--- a/Sources/AsyncHTTPClient/ConnectionsState.swift
+++ b/Sources/AsyncHTTPClient/ConnectionsState.swift
@@ -64,7 +64,7 @@ extension HTTP1ConnectionProvider {
         private var openedConnectionsCount: Int = 0
 
         /// Number of enqueued requests, used to track if it is safe to delete the provider.
-        private var pending: Int = 1
+        private var pending: Int = 0
 
         init(maximumConcurrentConnections: Int = 8, eventLoop: EventLoop) {
             self.maximumConcurrentConnections = maximumConcurrentConnections
@@ -148,7 +148,7 @@ extension HTTP1ConnectionProvider {
                     return .none
                 }
             case .closed:
-                return .fail(waiter, ProviderClosedError())
+                return .fail(waiter, HTTPClientError.alreadyShutdown)
             }
         }
 

--- a/Tests/AsyncHTTPClientTests/ConnectionPoolTests.swift
+++ b/Tests/AsyncHTTPClientTests/ConnectionPoolTests.swift
@@ -36,7 +36,7 @@ class ConnectionPoolTests: XCTestCase {
         XCTAssertEqual(0, snapshot.availableConnections.count)
         XCTAssertEqual(0, snapshot.leasedConnections.count)
         XCTAssertEqual(0, snapshot.waiters.count)
-        XCTAssertEqual(1, snapshot.pending)
+        XCTAssertEqual(0, snapshot.pending)
         XCTAssertEqual(0, snapshot.openedConnectionsCount)
 
         XCTAssertTrue(state.enqueue())
@@ -45,7 +45,7 @@ class ConnectionPoolTests: XCTestCase {
         XCTAssertEqual(0, snapshot.availableConnections.count)
         XCTAssertEqual(0, snapshot.leasedConnections.count)
         XCTAssertEqual(0, snapshot.waiters.count)
-        XCTAssertEqual(2, snapshot.pending)
+        XCTAssertEqual(1, snapshot.pending)
         XCTAssertEqual(0, snapshot.openedConnectionsCount)
     }
 
@@ -58,9 +58,10 @@ class ConnectionPoolTests: XCTestCase {
         XCTAssertEqual(0, snapshot.availableConnections.count)
         XCTAssertEqual(0, snapshot.leasedConnections.count)
         XCTAssertEqual(0, snapshot.waiters.count)
-        XCTAssertEqual(1, snapshot.pending)
+        XCTAssertEqual(0, snapshot.pending)
         XCTAssertEqual(0, snapshot.openedConnectionsCount)
 
+        XCTAssertTrue(state.enqueue())
         let action = state.acquire(waiter: .init(promise: self.eventLoop.makePromise(), setupComplete: self.eventLoop.makeSucceededFuture(()), preference: .indifferent))
         switch action {
         case .create(let waiter):
@@ -92,8 +93,10 @@ class ConnectionPoolTests: XCTestCase {
         XCTAssertEqual(1, snapshot.availableConnections.count)
         XCTAssertEqual(0, snapshot.leasedConnections.count)
         XCTAssertEqual(0, snapshot.waiters.count)
-        XCTAssertEqual(1, snapshot.pending)
+        XCTAssertEqual(0, snapshot.pending)
         XCTAssertEqual(1, snapshot.openedConnectionsCount)
+
+        XCTAssertTrue(self.http1ConnectionProvider.enqueue())
 
         let action = self.http1ConnectionProvider.state.acquire(waiter: .init(promise: self.eventLoop.makePromise(), setupComplete: self.eventLoop.makeSucceededFuture(()), preference: .indifferent))
         switch action {
@@ -127,8 +130,10 @@ class ConnectionPoolTests: XCTestCase {
         XCTAssertEqual(0, snapshot.availableConnections.count)
         XCTAssertEqual(0, snapshot.leasedConnections.count)
         XCTAssertEqual(0, snapshot.waiters.count)
-        XCTAssertEqual(1, snapshot.pending)
+        XCTAssertEqual(0, snapshot.pending)
         XCTAssertEqual(8, snapshot.openedConnectionsCount)
+
+        XCTAssertTrue(self.http1ConnectionProvider.enqueue())
 
         let action = self.http1ConnectionProvider.state.acquire(waiter: .init(promise: self.eventLoop.makePromise(), setupComplete: self.eventLoop.makeSucceededFuture(()), preference: .indifferent))
         switch action {
@@ -159,8 +164,10 @@ class ConnectionPoolTests: XCTestCase {
         XCTAssertEqual(0, snapshot.availableConnections.count)
         XCTAssertEqual(0, snapshot.leasedConnections.count)
         XCTAssertEqual(0, snapshot.waiters.count)
-        XCTAssertEqual(1, snapshot.pending)
+        XCTAssertEqual(0, snapshot.pending)
         XCTAssertEqual(0, snapshot.openedConnectionsCount)
+
+        XCTAssertTrue(state.enqueue())
 
         let action = state.acquire(waiter: .init(promise: self.eventLoop.makePromise(), setupComplete: self.eventLoop.makeSucceededFuture(()), preference: .delegateAndChannel(on: self.eventLoop)))
         switch action {
@@ -192,8 +199,10 @@ class ConnectionPoolTests: XCTestCase {
         XCTAssertEqual(1, snapshot.availableConnections.count)
         XCTAssertEqual(0, snapshot.leasedConnections.count)
         XCTAssertEqual(0, snapshot.waiters.count)
-        XCTAssertEqual(1, snapshot.pending)
+        XCTAssertEqual(0, snapshot.pending)
         XCTAssertEqual(1, snapshot.openedConnectionsCount)
+
+        XCTAssertTrue(self.http1ConnectionProvider.enqueue())
 
         let action = self.http1ConnectionProvider.state.acquire(waiter: .init(promise: channel.eventLoop.makePromise(), setupComplete: channel.eventLoop.makeSucceededFuture(()), preference: .delegateAndChannel(on: channel.eventLoop)))
         switch action {
@@ -231,8 +240,10 @@ class ConnectionPoolTests: XCTestCase {
         XCTAssertEqual(1, snapshot.availableConnections.count)
         XCTAssertEqual(0, snapshot.leasedConnections.count)
         XCTAssertEqual(0, snapshot.waiters.count)
-        XCTAssertEqual(1, snapshot.pending)
+        XCTAssertEqual(0, snapshot.pending)
         XCTAssertEqual(8, snapshot.openedConnectionsCount)
+
+        XCTAssertTrue(self.http1ConnectionProvider.enqueue())
 
         let action = self.http1ConnectionProvider.state.acquire(waiter: .init(promise: self.eventLoop.makePromise(), setupComplete: self.eventLoop.makeSucceededFuture(()), preference: .delegateAndChannel(on: self.eventLoop)))
         switch action {
@@ -266,8 +277,10 @@ class ConnectionPoolTests: XCTestCase {
         XCTAssertEqual(0, snapshot.availableConnections.count)
         XCTAssertEqual(0, snapshot.leasedConnections.count)
         XCTAssertEqual(0, snapshot.waiters.count)
-        XCTAssertEqual(1, snapshot.pending)
+        XCTAssertEqual(0, snapshot.pending)
         XCTAssertEqual(8, snapshot.openedConnectionsCount)
+
+        XCTAssertTrue(self.http1ConnectionProvider.enqueue())
 
         let action = self.http1ConnectionProvider.state.acquire(waiter: .init(promise: self.eventLoop.makePromise(), setupComplete: self.eventLoop.makeSucceededFuture(()), preference: .delegateAndChannel(on: self.eventLoop)))
         switch action {
@@ -1177,7 +1190,7 @@ class ConnectionPoolTests: XCTestCase {
         XCTAssertEqual(0, snapshot.availableConnections.count)
         XCTAssertEqual(1, snapshot.leasedConnections.count)
         XCTAssertEqual(0, snapshot.waiters.count)
-        XCTAssertEqual(1, snapshot.pending)
+        XCTAssertEqual(0, snapshot.pending)
         XCTAssertEqual(1, snapshot.openedConnectionsCount)
 
         connection.release(closing: false, logger: HTTPClient.loggingDisabled)
@@ -1187,7 +1200,7 @@ class ConnectionPoolTests: XCTestCase {
         XCTAssertEqual(1, snapshot.availableConnections.count)
         XCTAssertEqual(0, snapshot.leasedConnections.count)
         XCTAssertEqual(0, snapshot.waiters.count)
-        XCTAssertEqual(1, snapshot.pending)
+        XCTAssertEqual(0, snapshot.pending)
         XCTAssertEqual(1, snapshot.openedConnectionsCount)
 
         // cleanup
@@ -1211,7 +1224,7 @@ class ConnectionPoolTests: XCTestCase {
         XCTAssertEqual(0, snapshot.availableConnections.count)
         XCTAssertEqual(1, snapshot.leasedConnections.count)
         XCTAssertEqual(0, snapshot.waiters.count)
-        XCTAssertEqual(1, snapshot.pending)
+        XCTAssertEqual(0, snapshot.pending)
         XCTAssertEqual(1, snapshot.openedConnectionsCount)
 
         connection.release(closing: true, logger: HTTPClient.loggingDisabled)
@@ -1220,7 +1233,7 @@ class ConnectionPoolTests: XCTestCase {
         XCTAssertEqual(0, snapshot.availableConnections.count)
         XCTAssertEqual(0, snapshot.leasedConnections.count)
         XCTAssertEqual(0, snapshot.waiters.count)
-        XCTAssertEqual(1, snapshot.pending)
+        XCTAssertEqual(0, snapshot.pending)
         XCTAssertEqual(0, snapshot.openedConnectionsCount)
 
         // cleanup
@@ -1244,7 +1257,7 @@ class ConnectionPoolTests: XCTestCase {
         XCTAssertEqual(1, snapshot.availableConnections.count)
         XCTAssertEqual(0, snapshot.leasedConnections.count)
         XCTAssertEqual(0, snapshot.waiters.count)
-        XCTAssertEqual(1, snapshot.pending)
+        XCTAssertEqual(0, snapshot.pending)
         XCTAssertEqual(1, snapshot.openedConnectionsCount)
 
         connection.remoteClosed(logger: HTTPClient.loggingDisabled)
@@ -1253,7 +1266,7 @@ class ConnectionPoolTests: XCTestCase {
         XCTAssertEqual(0, snapshot.availableConnections.count)
         XCTAssertEqual(0, snapshot.leasedConnections.count)
         XCTAssertEqual(0, snapshot.waiters.count)
-        XCTAssertEqual(1, snapshot.pending)
+        XCTAssertEqual(0, snapshot.pending)
         XCTAssertEqual(0, snapshot.openedConnectionsCount)
 
         // cleanup
@@ -1277,7 +1290,7 @@ class ConnectionPoolTests: XCTestCase {
         XCTAssertEqual(1, snapshot.availableConnections.count)
         XCTAssertEqual(0, snapshot.leasedConnections.count)
         XCTAssertEqual(0, snapshot.waiters.count)
-        XCTAssertEqual(1, snapshot.pending)
+        XCTAssertEqual(0, snapshot.pending)
         XCTAssertEqual(1, snapshot.openedConnectionsCount)
 
         connection.timeout(logger: HTTPClient.loggingDisabled)
@@ -1286,7 +1299,7 @@ class ConnectionPoolTests: XCTestCase {
         XCTAssertEqual(0, snapshot.availableConnections.count)
         XCTAssertEqual(0, snapshot.leasedConnections.count)
         XCTAssertEqual(0, snapshot.waiters.count)
-        XCTAssertEqual(1, snapshot.pending)
+        XCTAssertEqual(0, snapshot.pending)
         XCTAssertEqual(0, snapshot.openedConnectionsCount)
 
         // cleanup
@@ -1309,8 +1322,10 @@ class ConnectionPoolTests: XCTestCase {
         XCTAssertEqual(1, snapshot.availableConnections.count)
         XCTAssertEqual(0, snapshot.leasedConnections.count)
         XCTAssertEqual(0, snapshot.waiters.count)
-        XCTAssertEqual(1, snapshot.pending)
+        XCTAssertEqual(0, snapshot.pending)
         XCTAssertEqual(1, snapshot.openedConnectionsCount)
+
+        XCTAssertTrue(self.http1ConnectionProvider.enqueue())
 
         let action = self.http1ConnectionProvider.state.acquire(waiter: .init(promise: self.eventLoop.makePromise(), setupComplete: self.eventLoop.makeSucceededFuture(()), preference: .indifferent))
         switch action {

--- a/Tests/AsyncHTTPClientTests/ConnectionPoolTests.swift
+++ b/Tests/AsyncHTTPClientTests/ConnectionPoolTests.swift
@@ -110,11 +110,10 @@ class ConnectionPoolTests: XCTestCase {
             XCTAssertEqual(0, snapshot.pending)
             XCTAssertEqual(1, snapshot.openedConnectionsCount)
 
-            // cleanup, since we don't call release
+            // cleanup
             // this cleanup code needs to go and use HTTP1ConnectionProvider's API instead
             // (https://github.com/swift-server/async-http-client/issues/234)
-            snapshot.openedConnectionsCount = 0
-            self.http1ConnectionProvider.state.testsOnly_setInternalState(snapshot)
+            self.http1ConnectionProvider.release(connection: connection, closing: true, logger: HTTPClient.loggingDisabled)
         default:
             XCTFail("Unexpected action: \(action)")
         }
@@ -216,11 +215,10 @@ class ConnectionPoolTests: XCTestCase {
             XCTAssertEqual(0, snapshot.pending)
             XCTAssertEqual(1, snapshot.openedConnectionsCount)
 
-            // cleanup, since we don't call release
+            // cleanup
             // this cleanup code needs to go and use HTTP1ConnectionProvider's API instead
             // (https://github.com/swift-server/async-http-client/issues/234)
-            snapshot.openedConnectionsCount = 0
-            self.http1ConnectionProvider.state.testsOnly_setInternalState(snapshot)
+            self.http1ConnectionProvider.release(connection: connection, closing: true, logger: HTTPClient.loggingDisabled)
         default:
             XCTFail("Unexpected action: \(action)")
         }
@@ -320,7 +318,6 @@ class ConnectionPoolTests: XCTestCase {
         default:
             XCTFail("Unexpected action: \(action)")
         }
-        print(state.testsOnly_getInternalState())
     }
 
     // MARK: - Release Tests
@@ -351,15 +348,14 @@ class ConnectionPoolTests: XCTestCase {
             XCTAssertEqual(0, snapshot.waiters.count)
             XCTAssertEqual(0, snapshot.pending)
             XCTAssertEqual(1, snapshot.openedConnectionsCount)
-
-            // cleanup
-            // this cleanup code needs to go and use HTTP1ConnectionProvider's API instead
-            // (https://github.com/swift-server/async-http-client/issues/234)
-            snapshot.openedConnectionsCount = 0
-            self.http1ConnectionProvider.state.testsOnly_setInternalState(snapshot)
         default:
             XCTFail("Unexpected action: \(action)")
         }
+
+        // cleanup
+        // this cleanup code needs to go and use HTTP1ConnectionProvider's API instead
+        // (https://github.com/swift-server/async-http-client/issues/234)
+        connection.remoteClosed(logger: HTTPClient.loggingDisabled)
     }
 
     func testReleaseAliveButClosingConnectionEmptyQueue() throws {
@@ -392,6 +388,11 @@ class ConnectionPoolTests: XCTestCase {
         default:
             XCTFail("Unexpected action: \(action)")
         }
+
+        // cleanup
+        // this cleanup code needs to go and use HTTP1ConnectionProvider's API instead
+        // (https://github.com/swift-server/async-http-client/issues/234)
+        self.http1ConnectionProvider.execute(action, logger: HTTPClient.loggingDisabled)
     }
 
     func testReleaseInactiveConnectionEmptyQueue() throws {
@@ -421,9 +422,15 @@ class ConnectionPoolTests: XCTestCase {
             XCTAssertEqual(0, snapshot.waiters.count)
             XCTAssertEqual(0, snapshot.pending)
             XCTAssertEqual(0, snapshot.openedConnectionsCount)
+
         default:
             XCTFail("Unexpected action: \(action)")
         }
+
+        // cleanup
+        // this cleanup code needs to go and use HTTP1ConnectionProvider's API instead
+        // (https://github.com/swift-server/async-http-client/issues/234)
+        self.http1ConnectionProvider.execute(action, logger: HTTPClient.loggingDisabled)
     }
 
     func testReleaseInactiveConnectionEmptyQueueHasConnections() throws {
@@ -455,15 +462,14 @@ class ConnectionPoolTests: XCTestCase {
             XCTAssertEqual(0, snapshot.waiters.count)
             XCTAssertEqual(0, snapshot.pending)
             XCTAssertEqual(1, snapshot.openedConnectionsCount)
-
-            // cleanup
-            // this cleanup code needs to go and use HTTP1ConnectionProvider's API instead
-            // (https://github.com/swift-server/async-http-client/issues/234)
-            snapshot.openedConnectionsCount = 0
-            self.http1ConnectionProvider.state.testsOnly_setInternalState(snapshot)
         default:
             XCTFail("Unexpected action: \(action)")
         }
+
+        // cleanup
+        // this cleanup code needs to go and use HTTP1ConnectionProvider's API instead
+        // (https://github.com/swift-server/async-http-client/issues/234)
+        connection.remoteClosed(logger: HTTPClient.loggingDisabled)
     }
 
     func testReleaseAliveConnectionHasWaiter() throws {
@@ -488,8 +494,8 @@ class ConnectionPoolTests: XCTestCase {
         let action = self.http1ConnectionProvider.state.release(connection: connection, closing: false)
         switch action {
         case .lease(let connection, let waiter):
-            // XCTAssertTrue(connection.isInUse)
             snapshot = self.http1ConnectionProvider.state.testsOnly_getInternalState()
+            XCTAssertTrue(snapshot.leasedConnections.contains(ConnectionKey(connection)))
             XCTAssertEqual(0, snapshot.availableConnections.count)
             XCTAssertEqual(1, snapshot.leasedConnections.count)
             XCTAssertEqual(0, snapshot.waiters.count)
@@ -500,8 +506,7 @@ class ConnectionPoolTests: XCTestCase {
             // this cleanup code needs to go and use HTTP1ConnectionProvider's API instead
             // (https://github.com/swift-server/async-http-client/issues/234)
             waiter.promise.succeed(connection)
-            snapshot.openedConnectionsCount = 0
-            self.http1ConnectionProvider.state.testsOnly_setInternalState(snapshot)
+            self.http1ConnectionProvider.release(connection: connection, closing: true, logger: HTTPClient.loggingDisabled)
         default:
             XCTFail("Unexpected action: \(action)")
         }
@@ -539,9 +544,8 @@ class ConnectionPoolTests: XCTestCase {
             // cleanup
             // this cleanup code needs to go and use HTTP1ConnectionProvider's API instead
             // (https://github.com/swift-server/async-http-client/issues/234)
-            waiter.promise.fail(TempError())
-            snapshot.openedConnectionsCount = 0
-            self.http1ConnectionProvider.state.testsOnly_setInternalState(snapshot)
+            // simulate create -> use -> release cycle
+            self.http1ConnectionProvider.connect(.failure(TempError()), waiter: waiter, logger: HTTPClient.loggingDisabled)
         default:
             XCTFail("Unexpected action: \(action)")
         }
@@ -580,9 +584,10 @@ class ConnectionPoolTests: XCTestCase {
             XCTAssertEqual(1, snapshot.openedConnectionsCount)
 
             // cleanup
+            // this cleanup code needs to go and use HTTP1ConnectionProvider's API instead
+            // (https://github.com/swift-server/async-http-client/issues/234)
             waiter.promise.succeed(connection)
-            snapshot.openedConnectionsCount = 0
-            self.http1ConnectionProvider.state.testsOnly_setInternalState(snapshot)
+            self.http1ConnectionProvider.release(connection: connection, closing: true, logger: HTTPClient.loggingDisabled)
         default:
             XCTFail("Unexpected action: \(action)")
         }
@@ -621,9 +626,10 @@ class ConnectionPoolTests: XCTestCase {
             XCTAssertEqual(1, snapshot.openedConnectionsCount)
 
             // cleanup
+            // this cleanup code needs to go and use HTTP1ConnectionProvider's API instead
+            // (https://github.com/swift-server/async-http-client/issues/234)
             waiter.promise.succeed(connection)
-            snapshot.openedConnectionsCount = 0
-            self.http1ConnectionProvider.state.testsOnly_setInternalState(snapshot)
+            self.http1ConnectionProvider.release(connection: connection, closing: true, logger: HTTPClient.loggingDisabled)
         default:
             XCTFail("Unexpected action: \(action)")
         }
@@ -665,9 +671,11 @@ class ConnectionPoolTests: XCTestCase {
             XCTAssertEqual(2, snapshot.openedConnectionsCount)
 
             // cleanup
-            waiter.promise.succeed(connection)
-            snapshot.openedConnectionsCount = 0
-            self.http1ConnectionProvider.state.testsOnly_setInternalState(snapshot)
+            // this cleanup code needs to go and use HTTP1ConnectionProvider's API instead
+            // (https://github.com/swift-server/async-http-client/issues/234)
+            // simulate create -> use -> release cycle
+            self.http1ConnectionProvider.connect(.failure(TempError()), waiter: waiter, logger: HTTPClient.loggingDisabled)
+            connection.remoteClosed(logger: HTTPClient.loggingDisabled)
         default:
             XCTFail("Unexpected action: \(action)")
         }
@@ -710,8 +718,8 @@ class ConnectionPoolTests: XCTestCase {
 
             // cleanup
             waiter.promise.succeed(replacement)
-            snapshot.openedConnectionsCount = 0
-            self.http1ConnectionProvider.state.testsOnly_setInternalState(snapshot)
+            connection.remoteClosed(logger: HTTPClient.loggingDisabled)
+            self.http1ConnectionProvider.release(connection: replacement, closing: true, logger: HTTPClient.loggingDisabled)
         default:
             XCTFail("Unexpected action: \(action)")
         }
@@ -752,9 +760,14 @@ class ConnectionPoolTests: XCTestCase {
             XCTAssertEqual(8, snapshot.openedConnectionsCount)
 
             // cleanup
+            // this cleanup code needs to go and use HTTP1ConnectionProvider's API instead
+            // (https://github.com/swift-server/async-http-client/issues/234)
             waiter.promise.fail(TempError())
-            snapshot.openedConnectionsCount = 0
+            snapshot.openedConnectionsCount = 2
             self.http1ConnectionProvider.state.testsOnly_setInternalState(snapshot)
+
+            snapshot.availableConnections.forEach { $0.remoteClosed(logger: HTTPClient.loggingDisabled) }
+            self.http1ConnectionProvider.release(connection: connection, closing: true, logger: HTTPClient.loggingDisabled)
         default:
             XCTFail("Unexpected action: \(action)")
         }
@@ -796,9 +809,10 @@ class ConnectionPoolTests: XCTestCase {
             XCTAssertEqual(1, snapshot.openedConnectionsCount)
 
             // cleanup
+            // this cleanup code needs to go and use HTTP1ConnectionProvider's API instead
+            // (https://github.com/swift-server/async-http-client/issues/234)
             waiter.promise.succeed(connection)
-            snapshot.openedConnectionsCount = 0
-            self.http1ConnectionProvider.state.testsOnly_setInternalState(snapshot)
+            self.http1ConnectionProvider.release(connection: connection, closing: true, logger: HTTPClient.loggingDisabled)
         default:
             XCTFail("Unexpected action: \(action)")
         }
@@ -838,9 +852,10 @@ class ConnectionPoolTests: XCTestCase {
             XCTAssertEqual(2, snapshot.openedConnectionsCount)
 
             // cleanup
-            waiter.promise.fail(TempError())
-            snapshot.openedConnectionsCount = 0
-            self.http1ConnectionProvider.state.testsOnly_setInternalState(snapshot)
+            // this cleanup code needs to go and use HTTP1ConnectionProvider's API instead
+            // (https://github.com/swift-server/async-http-client/issues/234)s
+            self.http1ConnectionProvider.connect(.failure(TempError()), waiter: waiter, logger: HTTPClient.loggingDisabled)
+            snapshot.availableConnections.forEach { $0.remoteClosed(logger: HTTPClient.loggingDisabled) }
         default:
             XCTFail("Unexpected action: \(action)")
         }
@@ -872,6 +887,11 @@ class ConnectionPoolTests: XCTestCase {
         default:
             XCTFail("Unexpected action: \(action)")
         }
+
+        // cleanup
+        // this cleanup code needs to go and use HTTP1ConnectionProvider's API instead
+        // (https://github.com/swift-server/async-http-client/issues/234)
+        self.http1ConnectionProvider.execute(action, logger: HTTPClient.loggingDisabled)
     }
 
     func testNextWaiterEmptyQueueHasConnections() throws {
@@ -905,10 +925,7 @@ class ConnectionPoolTests: XCTestCase {
             // cleanup
             // this cleanup code needs to go and use HTTP1ConnectionProvider's API instead
             // (https://github.com/swift-server/async-http-client/issues/234)
-            XCTAssertNoThrow(try available.close().wait())
-            snapshot.availableConnections.removeAll()
-            snapshot.openedConnectionsCount = 0
-            self.http1ConnectionProvider.state.testsOnly_setInternalState(snapshot)
+            available.remoteClosed(logger: HTTPClient.loggingDisabled)
         default:
             XCTFail("Unexpected action: \(action)")
         }
@@ -948,9 +965,8 @@ class ConnectionPoolTests: XCTestCase {
             // cleanup
             // this cleanup code needs to go and use HTTP1ConnectionProvider's API instead
             // (https://github.com/swift-server/async-http-client/issues/234)
-            waiter.promise.succeed(connection)
-            snapshot.openedConnectionsCount = 0
-            self.http1ConnectionProvider.state.testsOnly_setInternalState(snapshot)
+            waiter.promise.fail(TempError())
+            self.http1ConnectionProvider.release(connection: connection, closing: true, logger: HTTPClient.loggingDisabled)
         default:
             XCTFail("Unexpected action: \(action)")
         }
@@ -979,7 +995,7 @@ class ConnectionPoolTests: XCTestCase {
         let action = self.http1ConnectionProvider.state.processNextWaiter()
         switch action {
         case .lease(let connection, let waiter):
-            var snapshot = self.http1ConnectionProvider.state.testsOnly_getInternalState()
+            let snapshot = self.http1ConnectionProvider.state.testsOnly_getInternalState()
             XCTAssertTrue(snapshot.leasedConnections.contains(ConnectionKey(connection)))
             XCTAssertEqual(0, snapshot.availableConnections.count)
             XCTAssertEqual(1, snapshot.leasedConnections.count)
@@ -990,9 +1006,8 @@ class ConnectionPoolTests: XCTestCase {
             // cleanup
             // this cleanup code needs to go and use HTTP1ConnectionProvider's API instead
             // (https://github.com/swift-server/async-http-client/issues/234)
-            waiter.promise.succeed(connection)
-            snapshot.openedConnectionsCount = 0
-            self.http1ConnectionProvider.state.testsOnly_setInternalState(snapshot)
+            waiter.promise.fail(TempError())
+            self.http1ConnectionProvider.release(connection: connection, closing: true, logger: HTTPClient.loggingDisabled)
         default:
             XCTFail("Unexpected action: \(action)")
         }
@@ -1031,9 +1046,9 @@ class ConnectionPoolTests: XCTestCase {
             // cleanup
             // this cleanup code needs to go and use HTTP1ConnectionProvider's API instead
             // (https://github.com/swift-server/async-http-client/issues/234)
-            waiter.promise.fail(TempError())
-            snapshot.openedConnectionsCount = 0
-            self.http1ConnectionProvider.state.testsOnly_setInternalState(snapshot)
+            // simulate create -> use -> release cycle
+            self.http1ConnectionProvider.connect(.failure(TempError()), waiter: waiter, logger: HTTPClient.loggingDisabled)
+            available.remoteClosed(logger: HTTPClient.loggingDisabled)
         default:
             XCTFail("Unexpected action: \(action)")
         }
@@ -1071,6 +1086,11 @@ class ConnectionPoolTests: XCTestCase {
         default:
             XCTFail("Unexpected action: \(action)")
         }
+
+        // cleanup
+        // this cleanup code needs to go and use HTTP1ConnectionProvider's API instead
+        // (https://github.com/swift-server/async-http-client/issues/234)
+        self.http1ConnectionProvider.release(connection: connection, closing: true, logger: HTTPClient.loggingDisabled)
     }
 
     func testTimeoutAvailableConnection() throws {
@@ -1109,6 +1129,11 @@ class ConnectionPoolTests: XCTestCase {
         default:
             XCTFail("Unexpected action: \(action)")
         }
+
+        // cleanup
+        // this cleanup code needs to go and use HTTP1ConnectionProvider's API instead
+        // (https://github.com/swift-server/async-http-client/issues/234)
+        self.http1ConnectionProvider.execute(action, logger: HTTPClient.loggingDisabled)
     }
 
     func testRemoteClosedLeasedConnection() throws {
@@ -1141,6 +1166,11 @@ class ConnectionPoolTests: XCTestCase {
         default:
             XCTFail("Unexpected action: \(action)")
         }
+
+        // cleanup
+        // this cleanup code needs to go and use HTTP1ConnectionProvider's API instead
+        // (https://github.com/swift-server/async-http-client/issues/234)
+        self.http1ConnectionProvider.release(connection: connection, closing: true, logger: HTTPClient.loggingDisabled)
     }
 
     func testRemoteClosedAvailableConnection() throws {
@@ -1173,6 +1203,11 @@ class ConnectionPoolTests: XCTestCase {
         default:
             XCTFail("Unexpected action: \(action)")
         }
+
+        // cleanup
+        // this cleanup code needs to go and use HTTP1ConnectionProvider's API instead
+        // (https://github.com/swift-server/async-http-client/issues/234)
+        self.http1ConnectionProvider.execute(action, logger: HTTPClient.loggingDisabled)
     }
 
     // MARK: - Connection Tests
@@ -1206,8 +1241,7 @@ class ConnectionPoolTests: XCTestCase {
         // cleanup
         // this cleanup code needs to go and use HTTP1ConnectionProvider's API instead
         // (https://github.com/swift-server/async-http-client/issues/234)
-        snapshot.pending = 0
-        self.http1ConnectionProvider.state.testsOnly_setInternalState(snapshot)
+        connection.remoteClosed(logger: HTTPClient.loggingDisabled)
     }
 
     func testConnectionReleaseInactive() throws {
@@ -1235,12 +1269,6 @@ class ConnectionPoolTests: XCTestCase {
         XCTAssertEqual(0, snapshot.waiters.count)
         XCTAssertEqual(0, snapshot.pending)
         XCTAssertEqual(0, snapshot.openedConnectionsCount)
-
-        // cleanup
-        // this cleanup code needs to go and use HTTP1ConnectionProvider's API instead
-        // (https://github.com/swift-server/async-http-client/issues/234)
-        snapshot.pending = 0
-        self.http1ConnectionProvider.state.testsOnly_setInternalState(snapshot)
     }
 
     func testConnectionRemoteCloseRelease() throws {
@@ -1268,12 +1296,6 @@ class ConnectionPoolTests: XCTestCase {
         XCTAssertEqual(0, snapshot.waiters.count)
         XCTAssertEqual(0, snapshot.pending)
         XCTAssertEqual(0, snapshot.openedConnectionsCount)
-
-        // cleanup
-        // this cleanup code needs to go and use HTTP1ConnectionProvider's API instead
-        // (https://github.com/swift-server/async-http-client/issues/234)
-        snapshot.pending = 0
-        self.http1ConnectionProvider.state.testsOnly_setInternalState(snapshot)
     }
 
     func testConnectionTimeoutRelease() throws {
@@ -1301,12 +1323,6 @@ class ConnectionPoolTests: XCTestCase {
         XCTAssertEqual(0, snapshot.waiters.count)
         XCTAssertEqual(0, snapshot.pending)
         XCTAssertEqual(0, snapshot.openedConnectionsCount)
-
-        // cleanup
-        // this cleanup code needs to go and use HTTP1ConnectionProvider's API instead
-        // (https://github.com/swift-server/async-http-client/issues/234)
-        snapshot.pending = 0
-        self.http1ConnectionProvider.state.testsOnly_setInternalState(snapshot)
     }
 
     func testAcquireAvailableBecomesUnavailable() throws {
@@ -1354,13 +1370,11 @@ class ConnectionPoolTests: XCTestCase {
             XCTAssertEqual(0, snapshot.pending)
             XCTAssertEqual(1, snapshot.openedConnectionsCount)
 
-            waiter.promise.succeed(connection)
-
             // cleanup
             // this cleanup code needs to go and use HTTP1ConnectionProvider's API instead
             // (https://github.com/swift-server/async-http-client/issues/234)
-            snapshot.openedConnectionsCount = 0
-            self.http1ConnectionProvider.state.testsOnly_setInternalState(snapshot)
+            waiter.promise.fail(TempError())
+            self.http1ConnectionProvider.release(connection: connection, closing: true, logger: HTTPClient.loggingDisabled)
         default:
             XCTFail("Unexpected action: \(action)")
         }
@@ -1381,15 +1395,6 @@ class ConnectionPoolTests: XCTestCase {
     override func tearDown() {
         XCTAssertNotNil(self.eventLoop)
         XCTAssertNotNil(self.http1ConnectionProvider)
-        /* BEGIN workaround for #232, this whole block is to be replaced by the commented out line below */
-        // not closing the provider here (https://github.com/swift-server/async-http-client/issues/232)
-        var state = self.http1ConnectionProvider.state.testsOnly_getInternalState()
-        if state.pending == 1, state.waiters.isEmpty, state.leasedConnections.isEmpty, state.openedConnectionsCount == 0 {
-            state.pending = 0
-            self.http1ConnectionProvider.state.testsOnly_setInternalState(state)
-        }
-        self.http1ConnectionProvider.closePromise.succeed(())
-        /* END workaround for #232 */
         XCTAssertNoThrow(try self.http1ConnectionProvider.close().wait())
         XCTAssertNoThrow(try self.eventLoop.syncShutdownGracefully())
         self.eventLoop = nil

